sql/lease.go

```diff
@@ -0,0 +1,500 @@
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/security"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/cockroachdb/cockroach/util/retry"
+	"github.com/gogo/protobuf/proto"
+)
+
+// TODO(pmattis): Periodically renew leases for tables that were used recently and
+// for which the lease will expire soon.
+
+const (
+	leaseDuration    = int64(5 * time.Minute)
+	minLeaseDuration = int64(time.Minute)
+)
+
+var (
+	errLeaseVersionChanged = errors.New("lease version changed")
+)
+
+func nanosToTime(nanos int64) time.Time {
+	return time.Unix(0, nanos)
+}
+
+func nanosToDTimestamp(nanos int64) parser.DTimestamp {
+	return parser.DTimestamp{Time: nanosToTime(nanos)}
+}
+
+// LeaseState holds the state for a lease. Exported only for testing.
+type LeaseState struct {
+	TableDescriptor
+	expiration int64
+	refcount   int
+}
+
+func (s *LeaseState) String() string {
+	return fmt.Sprintf("%d:%d", s.Version, s.expiration)
+}
+
+// Expiration returns the expiration time of the lease.
+func (s *LeaseState) Expiration() time.Time {
+	return nanosToTime(s.expiration)
+}
+
+// Refcount returns the reference count of the lease.
+func (s *LeaseState) Refcount() int {
+	return s.refcount
+}
+
+// LeaseStore implements the operations for acquiring and releasing leases and
+// publishing a new version of a descriptor. Exported only for testing.
+type LeaseStore struct {
+	db     client.DB
+	clock  *hlc.Clock
+	nodeID uint32
+}
+
+// jitteredLeaseDuration returns a randomly jittered duration from the interval
+// [0.75 * leaseDuration, 1.25 * leaseDuration].
+func (s LeaseStore) jitteredLeaseDuration() int64 {
+	return int64(float64(leaseDuration) * (0.75 + 0.5*rand.Float64()))
+}
+
+// Acquire a lease on the most recent version of a table descriptor.
+func (s LeaseStore) Acquire(tableID ID, minVersion uint32) (*LeaseState, error) {
+	lease := &LeaseState{}
+	lease.expiration = s.clock.Now().WallTime + s.jitteredLeaseDuration()
+
+	err := s.db.Txn(func(txn *client.Txn) error {
+		p := planner{txn: txn, user: security.RootUser}
+		plan, err := p.query(
+			fmt.Sprintf(`SELECT descriptor FROM system.descriptor WHERE id = %d`, tableID))
+		if err != nil {
+			return err
+		}
+		if !plan.Next() {
+			return fmt.Errorf("table ID %d not found", tableID)
+		}
+		if err := proto.Unmarshal([]byte(plan.Values()[0].(parser.DBytes)), &lease.TableDescriptor); err != nil {
+			return err
+		}
+		if plan.Next() {
+			return fmt.Errorf("unexpected multiple results on SELECT")
+		}
+		if err := plan.Err(); err != nil {
+			return err
+		}
```

Comment:
looks like a win to me. I think you can probably abstract out the "expect exactly one result row" bit though